### PR TITLE
feat: add cc_asn1_library rule to generate .asn1 changes

### LINKED ISSUE #9715: bazel: Investigate how to generate files from .asn1 configurations with asn1c

### PR DESCRIPTION
**Pull request by themarwhal**
Signed-off-by: Marie Bremner <marwhal@fb.com>

## Summary
This PR adds a custom logic for Magma to generate ASN1 files and make a library out of it. 
All flags were copied from https://github.com/magma/magma/blob/4ad87780c543b05baaf6511e987a225e4b5e1a97/lte/gateway/c/core/oai/tasks/ngap/CMakeLists.txt

Some notes on how I ended up with this solution.

### Why genrule doesn't work
using genrule to run asn1c does not work as genrule requires an explicit list of output files. As ASN1C generates ~ 100 files, I did not want to go this route. (which would involve listing out every single output file)

### Why I chose to write a custom rule
I found two main ways that people seem to get around this. The first is zipping up the generated files and declaring that as an output. Wasn't exactly sure how those files ended up being used though. The other option is to create a custom rule that uses `ctx.actions.declare_directory`. (See https://github.com/envoyproxy/envoy/blob/main/bazel/envoy_build_system.bzl#L74) I ended up going with the second approach as I could find more examples on sourcegraph that used `ctx.actions.declare_directory`.

This thread helped me figure out how to use CCInfo - https://groups.google.com/g/bazel-discuss/c/iH_Id2sCTe4/m/BvuHW7PVBAAJ

## How to use the new library definition
```
# in any BUILD file
cc_asn1_library(
     name = "asn1_r16",
     asn1_file = "messages/asn1/r16/r16.asn1",
     prefix = "Ngap_",
 )

cc_library(
   name = "...",
   srcs = [...],
   hdrs = [...],
   deps = [":asn1_r16"],
}
```

## Pending
1. figure out if compiling asn1c with bazel is an option, in order to increase determinism of asn1c use and reduce environmental dependencies by Bazel.

## Test Plan
After running `bazel clean` (not the full clean with --expunge)
Screen Shot 2021-10-14 at 3 19 52 PM

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: experimental/bazel-base/Dockerfile [hunk 1]
```diff
@@ -84,6 +84,19 @@ RUN git clone --branch v4.3 https://github.com/mfontanini/libtins.git && \
     # symlink to /usr/lib to match Magma VM
     ln -s /usr/local/lib/libtins.so /usr/lib/libtins.so
 
+RUN apt-get install -y --no-install-recommends \ 
+    libtool
+
+#####  asn1c
+RUN git clone https://gitlab.eurecom.fr/oai/asn1c.git && \
```
Comment (themarwhal) [inline experimental/bazel-base/Dockerfile:61]: look into generating this with Bazel

**Comment by electronjoe** [comment]
I modified the PR Description text mildly for outstanding tasks - to `figure out if compiling asn1c with bazel is an option, in order to increase determinism of asn1c use and reduce environmental dependencies by Bazel.`

File: tools/build_defs/asn1c.bzl [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Custom library and macros to generate files with asn1c"""
+
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+def _get_dir_name(name):
+    # we need to postfix the directory name with .cc to trick Bazel into thinking this is a valid input
```
Comment (electronjoe) [inline tools/build_defs/asn1c.bzl:17]: Is there any GH Issue we can cite about this approach (e.g. future fix / feature to Bazel that enables what you need here)? Or was this found in Google Groups?

File: tools/build_defs/asn1c.bzl [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    return name + ".cc"
+
+def _get_substitution_commands(ctx, dir_path):
```
Comment (electronjoe) [inline tools/build_defs/asn1c.bzl]: In my read of this file, I'm unable to piece together effectively what the various `_get_` private functions are doing for you.  Would it be possible to give a top level strategy comment / description in the file? To explain what is broadly going on in this library? I know these are private, this is not for public use but for future editors or individuals who want to copy your work here for another generation task.  On that note you might highlight that other approaches are preferred if you know in advance your output file naming (either in comment in this file or in PR desc, maybe both).  I can just see this being copy-pasted in the future (maybe).

File: tools/build_defs/asn1c.bzl [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    Args:
+        name: the name of rule
+        asn1_file: relative path to the .asn1 file that will be passed to asn1c
+        prefix: value that is set to ASN1C_PREFIX
+    """
+    gen_name = name + "_genrule"
+
+    flags = "-pdu=all -fcompound-names -fno-include-deps -gen-PER"
+    substitutions = {
+        "18446744073709551615": "18446744073709551615u",
```
Comment (electronjoe) [inline tools/build_defs/asn1c.bzl:167]: Wondering what these substitutions are

**Review by electronjoe — APPROVED** [review]

File: tools/build_defs/asn1c.bzl (hunk 3 above)
Comment (themarwhal) [reply tools/build_defs/asn1c.bzl]: yeah top level comment for what this rule does would be useful to have.

the _get functions I made were to keep the main command construction cleaner looking as it involves a lot of string formatting. Maybe _construct_X_ commands is a better name?

Comment (electronjoe) [reply tools/build_defs/asn1c.bzl]: `_construct_` may help yeah.  And then some explanation of what you are constructing for (which I suppose is specific to asn1c compiler command line needs?).

File: tools/build_defs/asn1c.bzl (hunk 2 above)
Comment (themarwhal) [reply tools/build_defs/asn1c.bzl:17]: If I don't have the right extension, I just get an error message saying that the file has to have a certain extension. I added a comment with a related GH issue for more context.

File: tools/build_defs/asn1c.bzl (hunk 3 above)
Comment (themarwhal) [reply tools/build_defs/asn1c.bzl]: added a comment for each helper function and a bigger picture description in the main library definition

File: tools/build_defs/asn1c.bzl (hunk 4 above)
Comment (themarwhal) [reply tools/build_defs/asn1c.bzl:167]: I took it from here: https://github.com/magma/magma/blob/14c1cf643a61d576b3d24642e17ed3911d19210d/lte/gateway/c/core/oai/tasks/s1ap/CMakeLists.txt#L35

But I think it is to suppress warnings about the constant being too large.

Added a comment on it in the library def.

Comment (electronjoe) [reply tools/build_defs/asn1c.bzl:167]: Oh my lol. OK well good to know the origin at least.

**Review by electronjoe — APPROVED** [review]

**Review by ardzoht — APPROVED** [review]
lgtm

**Pull request merged**